### PR TITLE
[Merged by Bors] - feat: `positivity` extension for `posPart`, `negPart`

### PR DESCRIPTION
**Pull request by YaelDillies**
---

- [x] depends on: #10256

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 1]
```diff
@@ -415,3 +416,29 @@ def evalAscFactorial : PositivityExt where eval {u α} _ _ e := do
...
+  match e with
+  | ~q(@posPart _ $instαlat $instαgrp $a) =>
+    assertInstancesCommute
+    match ← core zα pα a with
+    | .positive pf => return .positive q(posPart_pos $pf)
+    | _ => return .nonnegative q(posPart_nonneg $a)
+  | _ => throwError "not `posPart`"
+
+/-- Extension for `negPart`. `a⁻` is always nonegative. -/
+@[positivity _⁺]
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Tactic/Positivity/Basic.lean]: Indeed!

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 2]
```diff
@@ -447,3 +448,29 @@ def evalRatDen : PositivityExt where eval {u α} _ _ e := do
...
     pure $ .positive q(den_pos $a)
   | _, _ => throwError "not Rat.num"
+
+/-- Extension for `posPart`. `a⁺` is always nonegative, and positive if `a` is. -/
+@[positivity _⁺]
+def evalPosPart : PositivityExt where eval zα pα e := do
+  match e with
+  | ~q(@posPart _ $instαlat $instαgrp $a) =>
+    assertInstancesCommute
+    match ← catchNone (core zα pα a) with
```
Comment (YaelDillies) [inline Mathlib/Tactic/Positivity/Basic.lean:458]: @eric-wieser, do you think this `catchNone` is a code smell? I would expect `core` not to fail but return `none`

Comment (eric-wieser) [reply Mathlib/Tactic/Positivity/Basic.lean:456]: I think the whole error handling model of positivity is a code smell...

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 3]
```diff
@@ -446,3 +447,29 @@ def evalRatDen : PositivityExt where eval {u α} _ _ e := do
...
+/-- Extension for `negPart`. `a⁻` is always nonegative. -/
+@[positivity _⁻]
+def evalNegPart : PositivityExt where eval _ _ e := do
+  match e with
+  | ~q(@negPart _ $instαlat $instαgrp $a) =>
+    assertInstancesCommute
+    return .nonnegative q(negPart_nonneg $a)
+  | _ => throwError "not `negPart`"
+
+-- There seems to be a bug in `Positivity.core` that makes it fail (instead of returning `.none`) in
```
Comment (dwrensha) [inline Mathlib/Tactic/Positivity/Basic.lean:471]: This comment appears to no longer apply.

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 4]
```diff
@@ -446,3 +447,29 @@ def evalRatDen : PositivityExt where eval {u α} _ _ e := do
...
+-- the first example
+example (a : ℤ) : 0 ≤ a⁺ := by positivity
```
Comment (dwrensha) [inline Mathlib/Tactic/Positivity/Basic.lean:473]: I think the convention is that these examples should go in `test/positivity.lean`.

**Comment by dwrensha** [comment]
maintainer merge

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**